File: sdks/go/http_wrap.go
package amlexia

import (
	"net/http"
	"strings"
	"time"
)

func providerFromURL(url string) string {
	if strings.Contains(url, "openai") {
		return "openai"
	}
	if strings.Contains(url, "anthropic") {
		return "anthropic"
	}
	if strings.Contains(url, "stripe") {
		return "stripe"
	}
	return ""
}

type trackingRoundTripper struct {
	base   http.RoundTripper
	client *Client
}

func (t *trackingRoundTripper) RoundTrip(req *http.Request) (*http.Response, error) {
	start := time.Now()
	status := 500
	var err error
	var res *http.Response
	res, err = t.base.RoundTrip(req)
	if res != nil {
		status = res.StatusCode
	}
	latency := int(time.Since(start).Milliseconds())
	prov := providerFromURL(req.URL.String())
	e := Event{
		Endpoint:   req.URL.String(),
		Method:     req.Method,
		StatusCode: status,
		LatencyMs:  latency,
		Provider:   &prov,
	}
	if err != nil {
		msg := err.Error()
		e.ErrorMessage = &msg
	}
	_ = t.client.Track(e)
	return res, err
}

// WrapHTTPClient returns an http.Client that tracks outbound requests.
func WrapHTTPClient(c *Client, base *http.Client) *http.Client {
	if base == nil {
		base = &http.Client{Timeout: 30 * time.Second}
	}
	rt := base.Transport
	if rt == nil {
		rt = http.DefaultTransport
	}
	clone := *base
	clone.Transport = &trackingRoundTripper{base: rt, client: c}
	return &clone
}

File: sdks/node/src/http-wrap.ts
import type { AmlexiaClient } from './client.js';
import type { TrackEvent } from './types.js';

export interface FetchWrapOptions {
  provider?: string;
  operationName?: string;
}

export function wrapFetch(
  client: AmlexiaClient,
  baseFetch: typeof fetch = fetch,
): typeof fetch {
  return async (input, init) => {
    const url = typeof input === 'string' ? input : input instanceof URL ? input.href : input.url;
    const method = (init?.method ?? 'GET').toUpperCase();
    const start = Date.now();
    let statusCode = 0;
    let errorMessage: string | undefined;
    try {
      const res = await baseFetch(input, init);
      statusCode = res.status;
      return res;
    } catch (err) {
      statusCode = 0;
      errorMessage = err instanceof Error ? err.message : String(err);
      throw err;
    } finally {
      client.track({
        endpoint: url,
        method,
        statusCode: statusCode || 500,
        latencyMs: Date.now() - start,
        errorMessage: errorMessage ?? null,
        provider: tryProviderFromUrl(url),
      });
    }
  };
}

function tryProviderFromUrl(url: string): string | null {
  try {
    const host = new URL(url).hostname;
    if (host.includes('openai')) return 'openai';
    if (host.includes('anthropic')) return 'anthropic';
    if (host.includes('stripe')) return 'stripe';
    return null;
  } catch {
    return null;
  }
}

export function trackHttpCall(
  client: AmlexiaClient,
  event: Omit<TrackEvent, 'latencyMs'> & { latencyMs?: number },
): void {
  client.track({
    ...event,
    latencyMs: event.latencyMs ?? 0,
  });
}
